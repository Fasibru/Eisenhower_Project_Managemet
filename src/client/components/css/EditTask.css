/* this class will occupy the whole screen so the user has to interact with the editTask popup
------------------------------------------------ */
.editTask-outer {
  position: fixed; /* necessary so the div around the popup will be placed at the top left */
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0, 0.3);
}

/* this class handles the styling of the newTask popup
------------------------------------------------ */
.editTask-inner {
  position: absolute;
  left: 25%;
  right: 25%;
  top: 25%;
  bottom: 25%;
  background: white;
}