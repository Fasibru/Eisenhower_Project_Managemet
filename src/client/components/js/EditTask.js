import React, { Component } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { connect } from 'react-redux';
import {
  closeEditTaskPopup,
  storeEditTaskFormChange,
  saveEditedTask,
  deleteTask,
} from '../../actions/index';
import '../scss/EditTask.scss';

const mapStateToProps = state => ({
  editTask: state.tasks.editTask,
  tasks: state.tasks.tasks,
});

export class EditTask extends Component {
  handleChange = (event) => {
    // eslint-disable-next-line no-shadow
    const { storeEditTaskFormChange } = this.props;

    // reflect form changes in editTask state
    if (event.target.name === 'completed') {
      storeEditTaskFormChange(event.target.name, event.target.checked);
    } else {
      storeEditTaskFormChange(event.target.name, event.target.value);
    }
  };

  handleSubmit = (event) => {
    const {
      editTask,
      tasks,
      // eslint-disable-next-line no-shadow
      saveEditedTask,
      // eslint-disable-next-line no-shadow
      closeEditTaskPopup,
    } = this.props;
    event.preventDefault();

    // save changes to the DB and if successful to UI as well
    axios.put(`/api/task/${editTask._id}`, editTask)
      .then(() => {
        // find index of task to update:
        const editTaskIndex = tasks.findIndex(task => task._id === editTask._id);

        // save changes in tasks array as well to avoid additional GET of all tasks
        saveEditedTask(editTask, editTaskIndex);

        // close the dialog after submit
        closeEditTaskPopup();
      })
      .catch((err) => {
        console.log(err);
        closeEditTaskPopup();
      });
  };

  handleDelete = () => {
    const {
      editTask,
      tasks,
      // eslint-disable-next-line no-shadow
      deleteTask,
      // eslint-disable-next-line no-shadow
      closeEditTaskPopup,
    } = this.props;

    axios.delete(`/api/task/${editTask._id}`)
      .then(() => {
        // remove deleted task from tasks array
        const deleteTaskIndex = tasks
          .findIndex(task => task._id === editTask._id);
        deleteTask(deleteTaskIndex);

        // close the popup after submit
        closeEditTaskPopup();
      })
      .catch((err) => {
        console.log(err);
        closeEditTaskPopup();
      });
  };

  render() {
    const {
      // eslint-disable-next-line no-shadow
      closeEditTaskPopup,
      // eslint-disable-next-line no-shadow
      editTask,
    } = this.props;

    return (
      <div className="editTask-outer">
        <div className="editTask-inner">
          <form onSubmit={this.handleSubmit}>
            <label htmlFor="title">
              Title<br />
              <input type="text" defaultValue={editTask.title} name="title" onChange={this.handleChange} />
            </label>
            <br />
            <label htmlFor="description">
              Description<br />
              <textarea type="text" name="description" defaultValue={editTask.description} onChange={this.handleChange} />
            </label>
            <br />
            <label htmlFor="category">
              Category<br />
              <input type="radio" name="category" value="A" onChange={this.handleChange} defaultChecked={editTask.category === 'A'} />A
              <input type="radio" name="category" value="B" onChange={this.handleChange} defaultChecked={editTask.category === 'B'} />B
              <input type="radio" name="category" value="C" onChange={this.handleChange} defaultChecked={editTask.category === 'C'} />C
              <input type="radio" name="category" value="D" onChange={this.handleChange} defaultChecked={editTask.category === 'D'} />D
            </label>
            <br />
            <input type="checkbox" name="completed" checked={editTask.completed} onChange={this.handleChange} />Completed
            <br />
            <input type="submit" value="Save" />
            <button type="button" onClick={closeEditTaskPopup}>Close</button>
            <button type="button" onClick={this.handleDelete}>Delete</button>
          </form>
        </div>
      </div>
    );
  }
}

EditTask.propTypes = {
  editTask: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    category: PropTypes.oneOf(['A', 'B', 'C', 'D']).isRequired,
    rank: PropTypes.number.isRequired,
    _id: PropTypes.string.isRequired,
  }).isRequired,
  closeEditTaskPopup: PropTypes.func.isRequired,
  storeEditTaskFormChange: PropTypes.func.isRequired,
  saveEditedTask: PropTypes.func.isRequired,
  deleteTask: PropTypes.func.isRequired,
  tasks: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default connect(mapStateToProps, {
  closeEditTaskPopup,
  storeEditTaskFormChange,
  saveEditedTask,
  deleteTask,
})(EditTask);
