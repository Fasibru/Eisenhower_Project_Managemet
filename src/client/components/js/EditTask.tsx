import axios from 'axios';
import * as React from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  closeEditTaskPopup,
  deleteTask,
  saveEditedTask,
  storeEditTaskFormChange,
} from '../../actions/actionsTasks';

import '../scss/EditTask.scss';
import '../scss/TaskForm.scss';

import TaskFormCategory from './TaskFormCategory';
import TaskFormDescription from './TaskFormDescription';
import TaskFormTitle from './TaskFormTitle';
import TaskWarningNote from './TaskWarningNote';

import { Filter, Store, TaskType } from '../../../types/storeTypes';
import { TaskActionsTypes } from '../../../types/taskActionTypes';

interface EditTaskProps {
  filters: Filter;
  editTask: TaskType;
  closeEditTaskPopup(): TaskActionsTypes;
  deleteTask(id: string): TaskActionsTypes;
  saveEditedTask(task: TaskType): TaskActionsTypes;
  storeEditTaskFormChange(name: string, value: boolean | string): TaskActionsTypes;
}

const mapStateToProps = (state: Store) => ({
  editTask: state.tasks.editTask,
  filters: state.filters.filters,
});

// tslint:disable-next-line: variable-name
export const EditTask: React.FC<EditTaskProps> = ({
  // tslint:disable: no-shadowed-variable
  editTask,
  filters,
  closeEditTaskPopup,
  deleteTask,
  saveEditedTask,
  storeEditTaskFormChange,
  // tslint:enable: no-shadowed-variable
}) => {
  const handleChange = (
    event: React.ChangeEvent<HTMLInputElement> | React.ChangeEvent<HTMLTextAreaElement>,
    ) => {
    // reflect form changes in editTask state
    if (event.target.name === 'completed') {
      storeEditTaskFormChange(event.target.name, (event.target as HTMLInputElement).checked);
    } else {
      storeEditTaskFormChange(event.target.name, event.target.value);
    }
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    // save changes to the DB and if successful to UI as well
    axios.put(`/api/task/${editTask._id}`, editTask)
      .then(() => {
        // save changes in tasks array for UI
        saveEditedTask(editTask);

        // close the dialog after submit
        closeEditTaskPopup();
      })
      .catch((err) => {
        console.log(err);
        closeEditTaskPopup();
      });
  };

  // delete task from the DB and if successful in UI as well
  const handleDelete = () => {
    axios.delete(`/api/task/${editTask._id}`)
      .then(() => {
        deleteTask(editTask._id);

        // close the popup after submit
        closeEditTaskPopup();
      })
      .catch((err) => {
        console.log(err);
        closeEditTaskPopup();
      });
  };

  const violatedFilters: string[] = [];
  if (editTask.description.indexOf(filters.searchQuery) === -1
    && editTask.title.indexOf(filters.searchQuery) === -1
  ) {
    violatedFilters.push(`Search tasks: ${filters.searchQuery}`);
  }
  if (editTask.completed && filters.showTasks === 'open') {
    violatedFilters.push('Show tasks: Open');
  }
  if (!editTask.completed && filters.showTasks === 'completed') {
    violatedFilters.push('Show tasks: Completed');
  }

  return (
    <div className="Task-outer">
      <div className="Task-inner">
        <p className="Task__header">Edit the Task</p>
        <form onSubmit={handleSubmit} className="Task__form">
          <TaskFormTitle
            title={editTask.title}
            handleChange={handleChange}
          />
          <TaskFormDescription
            description={editTask.description}
            handleChange={handleChange}
          />
          <TaskFormCategory
            category={editTask.category}
            handleChange={handleChange}
          />
          <input
            type="checkbox"
            className="Task__completed"
            name="completed"
            checked={editTask.completed}
            onChange={handleChange}
          />
          Completed
          <br />
          <input
            type="submit"
            value="Save Changes"
            className="Task__btn Task__btn--margin-right"
          />
          <button
            type="button"
            onClick={handleDelete}
            className="Task__btn Task__btn--margin-right"
          >
            Delete Task
          </button>
          <button
            type="button"
            onClick={closeEditTaskPopup}
            className="Task__btn"
          >
            Cancel
          </button>
          {
            violatedFilters.length > 0 && (
              <TaskWarningNote identifier="New task" violatedFilters={violatedFilters} />
            )
          }
        </form>
      </div>
    </div>
  );
};

// EditTask.propTypes = {
//   editTask: PropTypes.shape({
//     title: PropTypes.string.isRequired,
//     description: PropTypes.string.isRequired,
//     category: PropTypes.oneOf(['A', 'B', 'C', 'D']).isRequired,
//     rank: PropTypes.number.isRequired,
//     _id: PropTypes.string.isRequired,
//     completed: PropTypes.bool.isRequired,
//   }).isRequired,
//   closeEditTaskPopup: PropTypes.func.isRequired,
//   storeEditTaskFormChange: PropTypes.func.isRequired,
//   saveEditedTask: PropTypes.func.isRequired,
//   deleteTask: PropTypes.func.isRequired,
// };

export default connect(mapStateToProps, {
  closeEditTaskPopup,
  deleteTask,
  saveEditedTask,
  storeEditTaskFormChange,
})(EditTask);
