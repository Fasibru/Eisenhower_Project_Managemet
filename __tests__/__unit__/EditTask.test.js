import React from 'react';
import { shallow } from 'enzyme';
import EditTask from '../../src/client/components/js/EditTask';

describe('Testing <EditTask />', () => {
  const fnClick = jest.fn();
  const fnSubmit = jest.fn();
  const fnChange = jest.fn();
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<EditTask
      toggleEditTaskPopup={fnClick}
      editTask={{
        _id: '2',
        rank: -999,
        category: 'A',
        title: 'Task Title',
        description: 'Task Description',
        completed: false,
      }}
      defineEditTask={fnChange}
      submitEditTask={fnSubmit}
      handleDeleteTask={fnClick}
    />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should populate "title", "description", "category", "completed" correctly from props', () => {
    const inputTitle = wrapper.find('input[name="title"]');
    const textareaDescription = wrapper.find('textarea[name="description"]');
    const radioCategory = wrapper.find('input[type="radio"]');
    const checkboxCompleted = wrapper.find('input[type="checkbox"][name="completed"]');

    expect(inputTitle.props().defaultValue).toEqual('Task Title');
    expect(textareaDescription.props().defaultValue).toEqual('Task Description');
    expect(radioCategory.at(0).props().defaultChecked).toEqual(true);
    expect(radioCategory.at(1).props().defaultChecked).toEqual(false);
    expect(radioCategory.at(2).props().defaultChecked).toEqual(false);
    expect(radioCategory.at(3).props().defaultChecked).toEqual(false);
    expect(checkboxCompleted.props().checked).toEqual(false);
  });

  it('Should call "toggleEditTaskPopup" with "Close" button', () => {
    const buttonClose = wrapper.find('button').at(0);
    buttonClose.simulate('click');

    expect(fnClick).toHaveBeenCalledTimes(1);
  });

  it('Should call "handleDeleteTask" with "Delete" button', () => {
    const buttonDelete = wrapper.find('button').at(1);
    buttonDelete.simulate('click');

    expect(fnClick).toHaveBeenCalledTimes(1);
  });

  it('Should call "defineEditTask" on change', () => {
    const event = {
      target:
        { value: '' },
    };

    const eventChecked = {
      target:
        { checked: true },
    };

    const inputTitle = wrapper.find('input[name="title"]');
    inputTitle.simulate('change', event);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();

    const textareaDescription = wrapper.find('textarea[name="description"]');
    textareaDescription.simulate('change', event);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();

    const radioCategory = wrapper.find('input[type="radio"]');
    radioCategory.at(0).simulate('change', eventChecked);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();

    radioCategory.at(1).simulate('change', eventChecked);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();

    radioCategory.at(2).simulate('change', eventChecked);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();

    radioCategory.at(3).simulate('change', eventChecked);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();

    const checkboxCompleted = wrapper.find('input[type="checkbox"][name="completed"]');
    checkboxCompleted.simulate('change', eventChecked);
    expect(fnChange).toHaveBeenCalledTimes(1);
    fnChange.mockClear();
  });

  it('Handles switch statement correctly', () => {
    const caseA = shallow(<EditTask
      toggleEditTaskPopup={fnClick}
      editTask={{
        _id: '2',
        rank: -999,
        category: 'A',
        title: 'Task Title',
        description: 'Task Description',
        completed: false,
      }}
      defineEditTask={fnChange}
      submitEditTask={fnSubmit}
      handleDeleteTask={fnClick}
    />);

    expect(caseA.find('input[type="radio"][value="A"]').props().defaultChecked).toEqual(true);
    expect(caseA.find('input[type="radio"][value="B"]').props().defaultChecked).toEqual(false);
    expect(caseA.find('input[type="radio"][value="C"]').props().defaultChecked).toEqual(false);
    expect(caseA.find('input[type="radio"][value="D"]').props().defaultChecked).toEqual(false);

    const caseB = shallow(<EditTask
      toggleEditTaskPopup={fnClick}
      editTask={{
        _id: '2',
        rank: -999,
        category: 'B',
        title: 'Task Title',
        description: 'Task Description',
        completed: false,
      }}
      defineEditTask={fnChange}
      submitEditTask={fnSubmit}
      handleDeleteTask={fnClick}
    />);

    expect(caseB.find('input[type="radio"][value="A"]').props().defaultChecked).toEqual(false);
    expect(caseB.find('input[type="radio"][value="B"]').props().defaultChecked).toEqual(true);
    expect(caseB.find('input[type="radio"][value="C"]').props().defaultChecked).toEqual(false);
    expect(caseB.find('input[type="radio"][value="D"]').props().defaultChecked).toEqual(false);

    const caseC = shallow(<EditTask
      toggleEditTaskPopup={fnClick}
      editTask={{
        _id: '2',
        rank: -999,
        category: 'C',
        title: 'Task Title',
        description: 'Task Description',
        completed: false,
      }}
      defineEditTask={fnChange}
      submitEditTask={fnSubmit}
      handleDeleteTask={fnClick}
    />);

    expect(caseC.find('input[type="radio"][value="A"]').props().defaultChecked).toEqual(false);
    expect(caseC.find('input[type="radio"][value="B"]').props().defaultChecked).toEqual(false);
    expect(caseC.find('input[type="radio"][value="C"]').props().defaultChecked).toEqual(true);
    expect(caseC.find('input[type="radio"][value="D"]').props().defaultChecked).toEqual(false);

    const caseD = shallow(<EditTask
      toggleEditTaskPopup={fnClick}
      editTask={{
        _id: '2',
        rank: -999,
        category: 'D',
        title: 'Task Title',
        description: 'Task Description',
        completed: false,
      }}
      defineEditTask={fnChange}
      submitEditTask={fnSubmit}
      handleDeleteTask={fnClick}
    />);

    expect(caseD.find('input[type="radio"][value="A"]').props().defaultChecked).toEqual(false);
    expect(caseD.find('input[type="radio"][value="B"]').props().defaultChecked).toEqual(false);
    expect(caseD.find('input[type="radio"][value="C"]').props().defaultChecked).toEqual(false);
    expect(caseD.find('input[type="radio"][value="D"]').props().defaultChecked).toEqual(true);

    const caseDefault = shallow(<EditTask
      toggleEditTaskPopup={fnClick}
      editTask={{
        _id: '2',
        rank: -999,
        category: 'E',
        title: 'Task Title',
        description: 'Task Description',
        completed: false,
      }}
      defineEditTask={fnChange}
      submitEditTask={fnSubmit}
      handleDeleteTask={fnClick}
    />);

    expect(caseDefault.find('input[type="radio"][value="A"]').props().defaultChecked).toEqual(false);
    expect(caseDefault.find('input[type="radio"][value="B"]').props().defaultChecked).toEqual(false);
    expect(caseDefault.find('input[type="radio"][value="C"]').props().defaultChecked).toEqual(false);
    expect(caseDefault.find('input[type="radio"][value="D"]').props().defaultChecked).toEqual(false);
  });
});
