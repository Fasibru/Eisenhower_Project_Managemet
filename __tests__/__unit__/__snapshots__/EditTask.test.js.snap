// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <EditTask /> Renders correctly 1`] = `
<div
  className="editTask-outer"
>
  <div
    className="editTask-inner"
  >
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="title"
      >
        Title
        <br />
        <input
          defaultValue="Test title"
          name="title"
          onChange={[Function]}
          type="text"
        />
      </label>
      <br />
      <label
        htmlFor="description"
      >
        Description
        <br />
        <textarea
          defaultValue="Test description"
          name="description"
          onChange={[Function]}
          type="text"
        />
      </label>
      <br />
      <label
        htmlFor="category"
      >
        Category
        <br />
        <input
          defaultChecked={true}
          name="category"
          onChange={[Function]}
          type="radio"
          value="A"
        />
        A
        <input
          defaultChecked={false}
          name="category"
          onChange={[Function]}
          type="radio"
          value="B"
        />
        B
        <input
          defaultChecked={false}
          name="category"
          onChange={[Function]}
          type="radio"
          value="C"
        />
        C
        <input
          defaultChecked={false}
          name="category"
          onChange={[Function]}
          type="radio"
          value="D"
        />
        D
      </label>
      <br />
      <input
        checked={false}
        name="completed"
        onChange={[Function]}
        type="checkbox"
      />
      Completed
      <br />
      <input
        type="submit"
        value="Save"
      />
      <button
        onClick={[MockFunction]}
        type="button"
      >
        Close
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </form>
  </div>
</div>
`;
